File: hemin-engine/src/main/java/io/hemin/engine/model/FeedStatus.java
package io.hemin.engine.model;

import java.util.Collections;
import java.util.Map;
import java.util.concurrent.ConcurrentHashMap;

public enum FeedStatus {

    NEVER_CHECKED("never_checked"),
    DOWNLOAD_SUCCESS("download_success"),
    HTTP_403("http_403"),
    DOWNLOAD_ERROR("download_error"),
    PARSE_ERROR("parse_error");

    private String name;

    private static final Map<String,FeedStatus> STATUS_MAP;

    static {
        final Map<String,FeedStatus> map = new ConcurrentHashMap<>();
        for (FeedStatus instance : FeedStatus.values()) {
            map.put(instance.getName(),instance);
        }
        STATUS_MAP = Collections.unmodifiableMap(map);
    }

    private FeedStatus(String name){
        this.name = name;
    }

    public String getName(){
        return name;
    }

    public static FeedStatus getByName (String name) {
        return STATUS_MAP.get(name);
    }

}

File: hemin-engine/src/main/scala/io/hemin/engine/model/FeedStatus.scala
package io.hemin.engine.model

import enumeratum.{Enum, EnumEntry}

import scala.collection.immutable

sealed abstract class FeedStatus(override val entryName: String) extends EnumEntry

object FeedStatus extends Enum[FeedStatus] {

  /* `findValues` is a protected method that invokes a macro to
 * find all `FeedStatus` object declarations inside an `Enum`.
 * We use it to implement the `val values` member */
  val values: immutable.IndexedSeq[FeedStatus] = findValues

  final case object NeverChecked    extends FeedStatus("never_checked")
  final case object DownloadSuccess extends FeedStatus("download_success")
  final case object DownloadError   extends FeedStatus("download_error")
  final case object Http403         extends FeedStatus("http_403")
  final case object ParserError     extends FeedStatus("parser_error")

}
